src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -279,6 +324,99 @@ public bool TryRemove(TKey key, out TValue value)
             return false;
         }
 
+        /// <summary>
+        /// Remove an item from the LinkedList.
+        /// When the last item is removed from the LinkedList, the OnLinkedListItemRemoved() will cancel the running _eventQueue task.
+        /// </summary>
+        /// <param name="newCacheItem">the item to be removed</param>
+        private void RemoveItemFromLinkedList(LRUCacheItem<TKey, TValue> newCacheItem)
+        {
+            _doubleLinkedList.Remove(newCacheItem);
+            OnLinkedListItemRemoved();
+        }
+
+
+        /// <summary>
+        /// The method handling the event when an item is removed from the LinkedList.
+        /// The _eventQueueTask needs to be cancelled if the _doubleLinkedList is empty.
+        /// </summary>
+        private void OnLinkedListItemRemoved()
+        {
+            _eventQueueTaskState = EventQueueTaskStopRequested; // set to stopping to prevent task being started
+
+            // stop the tasks only when both the _doubleLinkedList and _eventQueue are empty, otherwise change back to running
+            if (_doubleLinkedList.Count != 0 || _eventQueue.Count != 0)
+            {
+                _eventQueueTaskState = EventQueueTaskRunning; // set to the state back to running
+            }
+        }
+
+        /// <summary>
+        /// The method handling the event when an item is added to the LinkedList.
+        /// The _eventQueueTask needs to be started if the _doubleLinkedList is empty.
+        /// </summary>
+        private void StartEventQueueTasksIfNotRunning()
+        {
+            // If the event queue task is set to be stopped but has not exited yet, abort the stopping process.
```

Comment:
abort the starting process?

Reply:
I added the following comments and hopefully that explains the logic better:
        /// This method is called after an item is added to the queue, so it needs to start the event queue task if it is not running (_eventQueueTaskState == EventQueueTaskRunning),
        /// and prevent the task from being stopped if it has been set to stop but not exited yet (the task while loop is still running).